UVicMartletplace/martletplace
# Gavin/212 add pgsync to keep elasticsearch up to date with postgres

# Description
I want to use [PGSync](https://pgsync.com/)  to keep elasticsearch in sync with Postgres (reading from postgres and writing to elasticsearch).


I was working in the martletplace repository on the search endpoint and I have realized a problem after chatting with Max.

It is very easy for the elasticsearch index which search depends upon to get out of sync with the Postgres database (the source of truth).

Originally, I had planned to forward Listings updates from the backend to the search endpoint using the /api/search/reindex/listing-created endpoint (and two others, which are present in the search endpoint)

Unfortunately, this has issues with staying in sync particularly when either service can potentially go down.


Closes #212

## How to Test
1. Run `docker-compose up -- build`

2. In a browser, paste `http://localhost/api/search?query=chocolate&minPrice=5&maxPrice=10&status=AVAILABLE&searchType=LISTINGS&latitude=48.4633&longitude=-123.3108&sort=RELEVANCE&page=1&limit=6`
expected output:
![image](https://github.com/UVicMartletplace/martletplace/assets/90437591/a660750f-5151-46d1-b35a-ef23e66614cd)

3. In a browser, paste `http://localhost/api/search?query=user&status=SOLD&searchType=USERS&latitude=48.4633&longitude=-123.3108&sort=RELEVANCE&page=1&limit=6`
expected output:
![image](https://github.com/UVicMartletplace/martletplace/assets/90437591/cd03f6b6-1878-4c3b-bf67-af6fbe4e141b)


## Checklist
